app/src/org/commcare/network/ModernHttpRequester.java

```diff
@@ -0,0 +1,224 @@
+                               boolean isPostRequest) {
+        this.isPostRequest = isPostRequest;
+        this.context = context;
+        this.params = params;
+        this.url = url;
+
+        setupAuthentication(isAuthenticatedRequest);
+    }
+
+    public void setResponseProcessor(HttpResponseProcessor responseProcessor) {
+        this.responseProcessor = responseProcessor;
+    }
+
+    private void setupAuthentication(boolean isAuth) {
+        if (isAuth) {
+            User u = getCurrentUser();
+            final String username = u.getUsername();
+            final String password = u.getCachedPwd();
+            if (username == null || password == null || User.TYPE_DEMO.equals(u.getUserType())) {
+                String message =
+                        "Trying to make authenticated http request without proper credentials";
+                throw new RuntimeException(message);
+            } else if (!"https".equals(url.getProtocol())) {
+                throw new PlainTextPasswordException();
+            } else {
+                // make authenticated requests
+                Authenticator.setDefault(new Authenticator() {
+                    @Override
+                    protected PasswordAuthentication getPasswordAuthentication() {
+                        return new PasswordAuthentication(
+                                HttpRequestGenerator.buildDomainUser(username),
+                                password.toCharArray());
+                    }
+                });
+            }
+        } else {
+            // clear any prior set authenticator to make unauthed requests
+            Authenticator.setDefault(null);
+        }
+    }
+
+    public static class PlainTextPasswordException extends RuntimeException {
+    }
+
+    private static User getCurrentUser() {
+        try {
+            return CommCareApplication._().getSession().getLoggedInUser();
+        } catch (SessionUnavailableException sue) {
+            throw new RuntimeException("Can't find user to make authenticated http request.");
+        }
+    }
+
+    public void request() {
+        try {
+            httpConnection = setupConnection(buildUrl());
+            processResponse();
+        } catch (IOException e) {
+            e.printStackTrace();
+            responseProcessor.handleIOException(e);
+        } finally {
+            if (httpConnection != null) {
+                httpConnection.disconnect();
+            }
+        }
+    }
+
+    private URL buildUrl() throws IOException {
+        if (isPostRequest) {
+            return url;
+        } else {
+            return buildUrlWithParams();
+        }
+    }
+
+    protected HttpURLConnection setupConnection(URL builtUrl) throws IOException {
+        Log.d(TAG, builtUrl.toString());
+        HttpURLConnection httpConnection = (HttpURLConnection)builtUrl.openConnection();
+        if (isPostRequest) {
+            setupConnectionInner(httpConnection);
```

Comment:
this line is repeated in the `else` statement, so can be moved to above the `if`